### PR TITLE
Adds async macro for easy async calls

### PR DESCRIPTION
**Pull request by cwfitzgerald**
There is still a bit of tooling to be desired right now, but this adds the first layer, giving full async callstacks, and easy to use async calls no matter the context.

### CODE REVIEW

File: src/async_utils/mod.rs [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+/// All functions called by this macro must take a [`Context`] as the first argument. The name of the argument does not
+/// matter. They must also all be `async`.
+///
+/// # Arguments:
+/// - `mode` is the type of invocation. This is one of the following identifiers: **REQUIRED**
+///   - Async Context:
+///     - `exec` invokes the function on the provided executor using [`spawn_with_handle`].
+///     - `inline` invokes the function directly.
+///   - Sync Context:
+///     - `from-sync` invokes the function on the provided executor using [`spawn_with_handle`]. It is the start of the
```
Comment (DethRaid) [inline src/async_utils/mod.rs]: I assume "sync" is short for "synchronous"? This doesn't seem like a particularly synchronous operation

File: src/async_utils/mod.rs [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+            file,
+            line,
+            column,
+            last: None,
+        })
+    }
+
+    #[doc(hidden)]
+    /// Append a new callstack. Only used by macro.
+    pub fn add_stack_frame(self: Arc<Self>, file: &'static str, line: u32, column: u32) -> Arc<StackFrame> {
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply src/async_utils/mod.rs]: Sure

File: src/async_utils/mod.rs (hunk 1 above)
Comment (cwfitzgerald) [reply src/async_utils/mod.rs]: Sync meaning synchronous meaning a non async function.

**Comment by cwfitzgerald** [comment]
Will I have time to work on this, only time will tell...

File: src/async_utils/mod.rs (hunk 1 above)
Comment (DethRaid) [reply src/async_utils/mod.rs]: Yeah I guess that's better for task composability

**Pull request merged**